src/index/DeltaTriples.h

```diff
@@ -0,0 +1,134 @@
+// Copyright 2023 - 2024, University of Freiburg
+//  Chair of Algorithms and Data Structures.
+//  Authors:
+//    2023 Hannah Bast <bast@cs.uni-freiburg.de>
+//    2024 Julian Mundhahs <mundhahj@tf.uni-freiburg.de>
+
+#pragma once
+
+#include "engine/LocalVocab.h"
+#include "global/IdTriple.h"
+#include "index/Index.h"
+#include "index/IndexBuilderTypes.h"
+#include "index/LocatedTriples.h"
+#include "parser/TurtleParser.h"
+#include "util/HashSet.h"
+
+// A class for maintaining triples that are inserted or deleted after index
+// building, we call these delta triples. How it works in principle:
+//
+// 1. For each delta triple, find the block index in each permutation (see
+// `LocatedTriple` in `index/LocatedTriples.h`).
+//
+// 2. For each permutation and each block, store a sorted list of the positions
+// of the delta triples within that block (see `LocatedTriplesPerBlock` in
+// `index/LocatedTriples.h`).
+//
+// 3. In the call of `PermutationImpl::scan`, use the respective lists to merge
+// the relevant delta triples into the index scan result.
+//
+// NOTE: The delta triples currently do not go well together with CACHING. See
+// the discussion at the end of this file.
+class DeltaTriples {
+ private:
+  // The index to which these triples are added.
+  const Index& index_;
+
+  // The local vocabulary of the delta triples (they may have components,
+  // which are not contained in the vocabulary of the original index).
+  LocalVocab localVocab_;
+
+  // The positions of the delta triples in each of the six permutations.
+  LocatedTriplesPerBlock locatedTriplesPerBlockInPSO_;
+  LocatedTriplesPerBlock locatedTriplesPerBlockInPOS_;
+  LocatedTriplesPerBlock locatedTriplesPerBlockInSPO_;
+  LocatedTriplesPerBlock locatedTriplesPerBlockInSOP_;
+  LocatedTriplesPerBlock locatedTriplesPerBlockInOSP_;
+  LocatedTriplesPerBlock locatedTriplesPerBlockInOPS_;
+
+  FRIEND_TEST(DeltaTriplesTest, insertTriplesAndDeleteTriples);
+
+  // Each delta triple needs to know where it is stored in each of the six
+  // `LocatedTriplesPerBlock` above.
+  struct LocatedTripleHandles {
+    LocatedTriples::iterator forPSO_;
+    LocatedTriples::iterator forPOS_;
+    LocatedTriples::iterator forSPO_;
+    LocatedTriples::iterator forSOP_;
+    LocatedTriples::iterator forOPS_;
+    LocatedTriples::iterator forOSP_;
+
+    LocatedTriples::iterator& forPermutation(Permutation::Enum permutation);
+  };
+
+  // The sets of triples added to and subtracted from the original index. In
+  // particular, no triple can be in both of these sets.
+  ad_utility::HashMap<IdTriple<0>, LocatedTripleHandles> triplesInserted_;
+  ad_utility::HashMap<IdTriple<0>, LocatedTripleHandles> triplesDeleted_;
+
+ public:
+  // Construct for given index.
+  explicit DeltaTriples(const Index& index) : index_(index) {}
+
+  // Get the common `LocalVocab` of the delta triples.
+  LocalVocab& localVocab() { return localVocab_; }
```

Comment:
It might not be, but it think that it be required anyways.
Mutable LocalVocabs are used in LocatedTriples and the execution of normal queries (though these are currently temporary). These two need to be in sync for the retrieval of IDs while querying, the temporary IDs while querying could be discarded. Given the constraints I decided on a single global LocalVocab.

Reply:
Yes, 
I think that issues like "how do we clean up the local vocab, when the inserted entries are not needed anymore"
Can be postponed to the future, as it is probably related with the question "how do we deal with concurrent updates" (in particular updates that run concurrently with queries.).